cuda_core/cuda/core/experimental/_memory.pyx

```diff
@@ -900,3 +903,499 @@ class _SynchronousMemoryResource(MemoryResource):
+
+        prop.location.type = VirtualMemoryResourceOptions._location_type_to_driver(config.location_type)
+        prop.location.id = self.device.device_id
+        prop.allocFlags.gpuDirectRDMACapable = 1 if config.gpu_direct_rdma else 0
+        prop.requestedHandleTypes = VirtualMemoryResourceOptions._handle_type_to_driver(config.handle_type)
+
+        # ---- Query and apply granularity ----
+        # Choose min vs recommended granularity per config
+        gran_flag = VirtualMemoryResourceOptions._granularity_to_driver(config.granularity)
+        res, gran = driver.cuMemGetAllocationGranularity(prop, gran_flag)
+        raise_if_driver_error(res)
+
+        aligned_size = self._align_up(size, gran)
+        addr_align = config.addr_align or gran
+
+        # ---- Transactional allocation ----
+        with Transaction() as trans:
+            # ---- Create physical memory ----
+            res, handle = driver.cuMemCreate(aligned_size, prop, 0)
+            raise_if_driver_error(res)
+            # Register undo for physical memory
+            trans.append(lambda: driver.cuMemRelease(handle))
+
+            # ---- Reserve VA space ----
+            # Potentially, use a separate size for the VA reservation from the physical allocation size
+            res, ptr = driver.cuMemAddressReserve(aligned_size, addr_align, config.addr_hint, 0)
+            raise_if_driver_error(res)
+            # Register undo for VA reservation
+            trans.append(lambda: driver.cuMemAddressFree(ptr, aligned_size))
+
+            # ---- Map physical memory into VA ----
+            res, = driver.cuMemMap(ptr, aligned_size, 0, handle, 0)
+            trans.append(lambda: driver.cuMemUnmap(ptr, aligned_size))
+            raise_if_driver_error(res)
+
+            # ---- Set access for owner + peers ----
+            descs = self._build_access_descriptors(prop)
+            if descs:
+                res, = driver.cuMemSetAccess(ptr, aligned_size, descs, len(descs))
+            trans.commit()
+
+        # Done — return a Buffer that tracks this VA range
+        buf = Buffer.from_handle(ptr=ptr, size=aligned_size, mr=self)
+        return buf
+
+    def deallocate(self, ptr: int, size: int, stream: Stream=None) -> None:
+        """
+        Deallocate memory on the device using CUDA VMM APIs.
+        """
+        result, handle = driver.cuMemRetainAllocationHandle(ptr)
+        raise_if_driver_error(result)
+        result, = driver.cuMemUnmap(ptr, size)
+        raise_if_driver_error(result)
+        result, = driver.cuMemAddressFree(ptr, size)
+        raise_if_driver_error(result)
+        result, = driver.cuMemRelease(handle)
+        raise_if_driver_error(result)
+
+
+    @property
+    def is_device_accessible(self) -> bool:
+        """
+        Indicates whether the allocated memory is accessible from the device.
+
+        Returns:
+            bool: Always True for NVSHMEM memory.
```

Comment:
This VMM memory resource isn't only for nvshmem though. Given we're only supporting device memory for now it still always returns True, but we may want to add a TODO comment to handle this based on `self.config.allocation_type`?

Reply:
This is a copy-paste legacy. Actually, it's probably easy enough to just support the hostalloc flag so we may as well do it here and I'll remove this entirely.